[SpeedDial] New component

mbrookes opened the pull request:
Closes #2851

Thought I would give this its first public airing for feedback and initial code review.

To do:

- ~~Complete the keyboard navigation story~~
- Tests (in progress)
- `.d.ts`

![kapture 2017-12-18 at 13 10 41](https://user-images.githubusercontent.com/357702/34107757-46ff16cc-e3f5-11e7-9aef-c22f1548c403.gif)

oliviertassinari commented on src/styles/zIndex.js:
```diff
@@ -12,6 +12,7 @@ const zIndex = {
   dialog: 1500,
   layer: 2000,
   popover: 2100,
+  speedDial: 2800,
```
2500 to be safer?

mbrookes replied:
Sure.

mbrookes commented:
On the keyboard navigation / accesibility question: I'm stuck as to whether up / down should follow the visible layout of the buttons, or follow the DOM layout:

```
<FAB>
<First Action>
<Second Action>
<Third Action>
```
This would behave more like a menu (down is next, up is previous), and might be more natural for a visually impaired user.

oliviertassinari commented on pages/demos/buttons.js:
```diff
@@ -29,6 +29,14 @@ module.exports = require('fs')
   .readFileSync(require.resolve('docs/src/pages/demos/buttons/FloatingActionButtons'), 'utf8')
 `,
         },
+
```
blank line

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,128 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { withStyles } from 'material-ui/styles';
+import classNames from 'classnames';
```
I try to sort the import by scope. Maybe you would want to do the same and to keep it before the material-ui imports.

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,128 @@
+    width: 56,
+    textAlign: 'center',
+    zIndex: theme.zIndex.speedDial,
+    transition: 'transform 0.1s ease-in-out',
```
Shouldn't we use our transition creator (for animation curve consistency)?

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,128 @@
+    transition: 'transform 0.1s ease-in-out',
+  },
+  hidden: {
+    transform: 'scale(0, 0)',
```
Does it need to be 2D?

oliviertassinari commented on src/SpeedDial/SpeedDialButton.js:
```diff
@@ -0,0 +1,77 @@
+
+  return (
+    <div className={classes.root} {...other}>
+      <Tooltip title={tooltip} placement="left">
```
We might want to add a `TooltipProps` and a `ButtonProps` property.

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,128 @@
+
+  const children = React.Children.map(childrenProp, (child, childIndex) => {
+    if (!React.isValidElement(child)) return null;
+    const delay = 30 * (open ? childrenProp.length - childIndex : childIndex);
```
Humm. I would rather be incrementing a counter to account for the null children that can be provided.

mbrookes replied:
Could you explain a bit?

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,128 @@
+  const rootClassName = classNames(
+    classes.root,
+    hidden && classes.hidden,
+    hidden && classes.hidden,
```
twice the same line.

oliviertassinari commented:
Did you benchmark any other implementations?

mbrookes commented:
> Did you benchmark any other implementations?

Not sure what you mean by benchmark. I took a look, they mostly suck.

mbrookes commented on docs/src/pages/demos/buttons/SpeedDials.js:
```diff
@@ -0,0 +1,83 @@
+import PropTypes from 'prop-types';
+import { withStyles } from 'material-ui/styles';
+import Button from 'material-ui/Button';
+import SpeedDialButton from 'material-ui/SpeedDial/SpeedDialButton';
```
Sure, if that's agreed. (I'm supporting it, but lets give others a chance to have a say).

leMaik replied:
I like the way it's exported below, so it can be imported like:

```js
import SpeedDial, { SpeedDialButton } from 'material-ui/SpeedDial';
```

This still groups imports by the component they're used with.

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,144 @@
+// @flow weak
```
You can remove flow

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,144 @@
+import { withStyles } from 'material-ui/styles';
+import Button from 'material-ui/Button';
+import Icon from 'material-ui/Icon';
+import AddIcon from 'material-ui-icons/Add';
```
No, we don't rely on `material-ui-icons`. It needs to be an internal svg icons. We already have ~6 of them.

mbrookes replied:
Oops, I should know that! (SpeedDial started as a docs demo, and grew from there, so was using originally material-ui-icons)

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,144 @@
+
+  const children = React.Children.map(childrenProp, (child, childIndex) => {
+    if (!React.isValidElement(child)) return null;
+    const delay = 30 * (open ? childIndex : childrenProp.length - childIndex);
```
My point was that the index can jump from 0, 1 to 3,4,5 if we have two null children along the way.
Hence, the delay will be wrong.

mbrookes replied:
Struggling to imagine a situation where that would be the case? If somone's using other than SpeedDialButton, that's an error.

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,144 @@
+  return (
+    <div className={rootClassName} {...other}>
+      <Button
+        id={label}
```
Humm. I'm wondering about id escaping here. `string -> id` might not be as straightforward.

mbrookes replied:
Mmm... good point. I originally had a separate prop, but was trying to keep the API surface down. I've added a regex to filter the label.

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,144 @@
+  /**
+   * If `true`, the SpeedDial is open.
+   */
+  open: PropTypes.bool,
```
Should it be required?

mbrookes replied:
Makes sense.

oliviertassinari commented on src/SpeedDial/SpeedDialButton.js:
```diff
@@ -0,0 +1,99 @@
+// @flow weak
```
You can remove flow

mbrookes replied:
Oops, thought I had done. Done.

oliviertassinari commented on test/integration/MenuList.spec.js:
```diff
@@ -4,8 +4,8 @@ import React from 'react';
-import MenuList from 'src/Menu/MenuList';
-import MenuItem from 'src/Menu/MenuItem';
+import MenuList from 'src/MenuList/MenuList';
+import MenuItem from 'src/MenuItem/MenuItem';
```
Wait, what?

mbrookes replied:
Bah, sorry, stray change from the `flat-src-dir` experiment. I need to be more careful with what `git add .` adds (or doesn't - I see how that got missed from the other commit...)

oliviertassinari commented:
@mbrooks This is true benchmarking 💃 !

mbrookes commented:
@oliviertassinari I've added up/down arrow key support. I solved the accessibility dilema (should up/down follow the visual presentation, or the DOM structure) by supporting both.

Whichever key ('up', or 'down') is used to move focus from the base FAB to the first action is thereafter the "next" key, with the opposite focusing the previous action, (or the base FAB if there is no previous action). 

It's pretty intuitive in use, and worth the extra few lines of code, and use of state IMHO.

Additional review feedback is also fixed (I hope!) in a657b7b .

Regarding the API, are you happy with `SpeedDialButton` (a la `BottomNavigationButton`), or would `SpeedDialItem` (like `ListItem` and `MenuItem`) be preferable, (or even `SpeedDialAction`?).

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,198 @@
+const styles = theme => ({
+  root: {
+    width: 56,
+    textAlign: 'center',
```
I don't think that it's needed.

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,198 @@
+    flexDirection: 'column-reverse', // Display the first action at the bottom.
+    marginBottom: 16,
+  },
+  closedActions: {
```
For consistency with  the other classes names on the codebase: `actionsClosed`

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,198 @@
+      duration: theme.transitions.duration.shortest,
+    }),
+  },
+  hiddenFab: {
```
`fabHidden`

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,198 @@
+      easing: theme.transitions.easing.easeInOut,
+    }),
+  },
+  openFabIcon: {
```
`fabIconOpen`

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,198 @@
+    prevKey: 'down',
+  };
+
+  handleKeyDown = (event: SyntheticUIEvent<>) => {
```
Maybe we can extract and share the same logic than with the MenuList. Right now, something is off. It doesn't loop correctly. I can get into a situation where the up and down arrow keys have a reverse effect.

mbrookes replied:
That is by design. See: https://github.com/mui-org/material-ui/pull/9539#issuecomment-353171073

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,198 @@
+    const actionsClassName = classNames(classes.actions, { [classes.closedActions]: !open });
+    const children = React.Children.map(childrenProp, (child, childIndex) => {
+      if (!React.isValidElement(child)) return null;
+      const delay = 30 * (open ? childIndex : childrenProp.length - childIndex);
```
Missing index counter to counterbalance the potential null children.

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,198 @@
+        <div
+          id={`${id}-actions`}
+          className={actionsClassName}
+          aria-labelledby={id}
```
From what I can see on the Inbox implementation, the aria-labeledby should be targeting an element with some text. Here. I think that you can remove this property and the regexp that comes with it.

mbrookes replied:
Inbox might be wrong. I have followed https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-links.html

mbrookes replied:
I may be mistaken, but I understood that this with in turn would respect the label applied with `aria-label` to the element referenced by id.

mbrookes replied:
@oliviertassinari Reading around, I think you're right. I've removed it.

oliviertassinari commented on src/SpeedDial/SpeedDialButton.js:
```diff
@@ -0,0 +1,130 @@
+  root: {
+    position: 'relative',
+  },
+  actionButton: {
```
maybe `button` is enough.

oliviertassinari commented on src/SpeedDial/SpeedDialButton.js:
```diff
@@ -0,0 +1,130 @@
+    opacity: 1,
+    color: theme.palette.text.secondary,
+  },
+  closedActionButton: {
```
`buttonClosed`

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,198 @@
+    }),
+  },
+  hiddenFab: {
+    transform: 'scale3d(0, 0, 1)',
```
Why using scale 3d and not?
```jsx
transform: 'scale(0)',
```

mbrookes replied:
Bacause I used scale, and you asked why I wasn't using scale3d. LOL!

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -101,9 +101,17 @@ class SpeedDial extends React.Component {
-    const children = React.Children.map(childrenProp, (child, childIndex) => {
+
+    let totalValidChildren = 0;
+    React.Children.forEach(childrenProp, child => {
```
Why a double loop?

mbrookes commented:
@oliviertassinari I've made an attempt to deal with null children. It required two counters, as we need the total number of valid children to calculate the reverse timing for the closing animation. I'm wide-open to alternative implementations.

oliviertassinari commented on src/svg-icons/Add.js:
```diff
@@ -0,0 +1,16 @@
+/**
+ * @ignore - internal component.
+ */
+let Add = props => (
```
To move with #9601

mbrookes replied:
👍

mbrookes commented:
@oliviertassinari Should I proxy all of these to the Button component of SpeedDial and SpeedDialButton?

```
      onBlur,
      onFocus,
      onKeyboardFocus,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseLeave,
      onMouseUp,
      onTouchEnd,
      onTouchMove,
      onTouchStart,
```

leMaik commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,221 @@
+        icon = <Icon>{iconProp}</Icon>;
+      }
+    } else {
+      icon = <AddIcon className={fabIconClassName} />;
```
It would be nice to be able to use a different icon when hovered (e.g. a plus when not hovered and a pen icon when hovered, like in Google Inbox). Maybe by adding an `activeIcon` prop? :thinking:

mbrookes replied:
Good idea.

leMaik commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,221 @@
+    const actionsClassName = classNames(classes.actions, { [classes.actionsClosed]: !open });
+
+    // Filter the label for valid id characters.
+    const id = label.replace(/^[^a-z]+|[^\w:.-]+/gi, '');
```
So... a unique label is required?

mbrookes replied:
Yes. The alternative was an explicit `id` prop, but I was trying to minimize the API surface. Given that you'd only ever have one SpeedDial visible, and if there's more than one they presumably have different purposes, it didn't seem like a major limitation, but I should probably add a note to the prop docs.

leMaik commented on src/internal/svg-icons/Add.js:
```diff
@@ -0,0 +1,16 @@
+Add = pure(Add);
+Add.muiName = 'SvgIcon';
+
+export default Add;
```
Unrelated to this PR, but would it be possible to import those icons from `material-ui-icons` instead of copying them?

mbrookes replied:
We'd end up with a circular dependancy (material-ui-icons depends on material-uo/SvgIcon), so not sure how that would work out in practice.

leMaik replied:
Oh, you're right. I guess this would be possible by `npm link`ing `packages/material-ui` into `packages/material-ui-icons` (lerna/lerna might even do this automagically, but I'm not sure). Maybe we should keep it as is then.

oliviertassinari replied:
yarn workspace is another option.

mbrookes commented:
Thanks for the review. Not sure why, but I can't reply to this comment:

> it can be imported like:
> `import SpeedDial, { SpeedDialButton } from 'material-ui/SpeedDial';`

Yes, I was importing separately in anticipation of #9532.

mbrookes commented:
@leMaik `openIcon` is done:

![kapture 2017-12-28 at 2 39 57](https://user-images.githubusercontent.com/357702/34398661-921c1aee-eb78-11e7-9c78-d7bec34f7451.gif)

(Sorry, don't know why the gif is double size!)

oliviertassinari commented on src/SpeedDial/SpeedDial.js:
```diff
@@ -0,0 +1,239 @@
+            data-mui-test="SpeedDialButton"
+            {...ButtonProps}
+          >
+            <SpeedDialIcon icon={icon} openIcon={openIcon} open={open} {...SpeedDialIconProps} />
```
What about reversing the abstraction: exposing `SpeedDialIcon` in the demos instead of the `SpeedDialIconProps` property? So people keep full control over what they add.

mbrookes replied:
Okay.

oliviertassinari commented on docs/src/pages/demos/buttons/buttons.md:
```diff
@@ -1,5 +1,5 @@
 ---
-components: Button, IconButton, ButtonBase
+components: Button, IconButton, ButtonBase, SpeedDial, SpeedDialButton
```
`SpeedDialIcon`?

mbrookes commented:
> I'm still facing the same keyboard issue. Here is me pressing the down array key multiple times

Ah, thanks for the screenshot. I had misunderstood what you meant previously.

mbrookes commented:
@oliviertassinari I can't reporoduce the keyboard issue. If I keep hitting down, it stops at the delete icon.

mbrookes commented:
Regarding down = up: as I mentioned when you raised this concern before, this is by design. Please see: https://github.com/mui-org/material-ui/pull/9539#issuecomment-353171073

When I saw your screenshot I thought you were now saying it loops from top to bottom, but just realise when I couldn't reproduce it that it's just your gif looping! (How are you taking a screenshot that doesn't inclue the time it takes to stop the recording? 😄)

mbrookes commented:
@oliviertassinari Regarding the API, are you happy with `SpeedDialButton` (a la `BottomNavigationButton`), or would `SpeedDialItem` (like `ListItem` and `MenuItem`) be preferable (Speed Dial is a sort of menu)? Or perhaps `SpeedDialAction`, since the MD spec calls these actions?

mbrookes commented:
Some tests have randomly started failing:

```
  2) <SpeedDialButton />
       should render a Tooltip:

      AssertionError: expected 'WithStyles' to equal 'withStyles(Tooltip)'
      + expected - actual

      -WithStyles
      +withStyles(Tooltip)
```
🤷‍♂️

oliviertassinari commented:
@mbrookes The material specification is using the "action" wording on the bottom navigation page too. I would vote for `BottomNavigationAction` and `SpeedDialAction `.

mbrookes commented:
Cool, thanks! I was leaning towards Action for SpeedDial, so great that BottomNav is the same.  I'll change that in a separate PR.

gregnb commented:
I was looking at what components still needed to be done for v1 and I saw that the Floating action Button - Toolbar: https://material.io/guidelines/components/buttons-floating-action-button.html#buttons-floating-action-button-transitions

I didn't want to try to work on something that was being done so I searched and came across this thread.  Now I know it's not the same exact component, but I was wondering though, since these all fold under FOB what if it had one general component to call that you could apply which type it was? 

```
  <ActionButton type="Toolbar|SpeedDial|Morph|Full screen">
    ...
  </ActionButton>
   
  or

  <FloatingAction type="Toolbar|SpeedDial|Morph|Full screen">
    ...
  </FloatingAction>

```

Or, is it better to just keep these separate? If so, I would like to work on that Toolbar one if it's ok

mbrookes commented:
@gregnb Thanks for raising this.

I could very easily have missed one (it's not like I've checked every app, so lease correct me here!), but I'm not aware of any applications that use the FAB Toolbar approach.

gregnb commented:
@mbrookes Ah ok, I will have to defer to you on this one. I really have not done the research on the FAB Toolbar usage. Was just browsing what was on the todo list for V1 :)

mbrookes commented:
Todo list?

gregnb commented:
Sorry, this list here of what is supported and what is not:
https://material-ui-next.com/getting-started/supported-components/

mbrookes commented:
Ah, right. We don't aim to (directly) support every use case, that's just to show what *is* supported.

RuneKR commented:
any news on when this gets merged?

sambegin commented:
post v1 milestone just been added to this issue. That means we won't see it released any soon ? Any help we could provide to accelerate the process of releasing this feature ?

oliviertassinari commented:
> post v1 milestone just been added to this issue. 

@sambegin I believe we can defer the work on this enhancement. It doesn't mean @mbrookes won't push it forward.

mbrookes commented:
@oliviertassinari Shall I open a PR for the WIP lab package?

oliviertassinari commented:
@mbrookes Sure :)

mbrookes commented:
Closed for #10288.

raymondsze commented:
I grabbed the pull request and used in my project. Seems there is performance issue in **Safari**.
When the SpeedDial button is clicked, only the first SpeedDialAction is shown. The second and later SpeedDialAction show after a long delay.

oliviertassinari commented:
@raymondsze The component was pushed on npm under `@material-ui/labl/SpeedDial`.

mbrookes commented:
@raymondsze I'm looking at [docs demo](https://material-ui-next.com/lab/speed-dial/) in Safari, and not seeing an issue. I'd suggest using the published version as @oliviertassinari mentioned.
